File: InterviewPreparationKit/Graphs/ShortestReach.cpp
// https://www.hackerrank.com/challenges/ctci-bfs-shortest-reach/problem
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <deque>

using namespace std;

class Graph {
    private:
        class Node {
            public:
                vector<int> neighbors;
        };
        vector<Node> nodes;
        int edge_weight = 6;
        
    public:
        Graph(int n) {
            for (int i = 0; i < n; i++) {
                nodes.insert(nodes.end(), Node());
            }
        }
    
        void add_edge(int u, int v) {
            nodes[u].neighbors.insert(nodes[u].neighbors.end(), v);
            nodes[v].neighbors.insert(nodes[v].neighbors.end(), u);
        }
    
        vector<int> shortest_reach(int start) {
            deque<int> queue;
            queue.push_back(start);
            
            vector<int> distances(nodes.size(), -1);
            distances[start] = 0;
            
            while (!queue.empty()) {
                int node = queue.front();
                queue.pop_front();
                for (int neighbor : nodes[node].neighbors) {
                    if (distances[neighbor] == -1) {
                        distances[neighbor] = distances[node] + edge_weight;
                        queue.push_back(neighbor);
                    }
                }
            }
            
            return distances;
        }
};

int main() {
    int queries;
    cin >> queries;
    
    for (int t = 0; t < queries; t++) {
      
        int n, m;
        cin >> n;
        // Create a graph of size n where each edge weight is 6: 
        Graph graph(n);
        cin >> m;
        // read and set edges
        for (int i = 0; i < m; i++) {
            int u, v;
            cin >> u >> v;
            u--, v--;
            // add each edge to the graph
            graph.add_edge(u, v);
        }
        int startId;
        cin >> startId;
        startId--;
        // Find shortest reach from node s
        vector<int> distances = graph.shortest_reach(startId);

        for (int i = 0; i < distances.size(); i++) {
            if (i != startId) {
                cout << distances[i] << " ";
            }
        }
        cout << endl;
    }
    
    return 0;
}

File: InterviewPreparationKit/Graphs/ShortestReach.py
# https://www.hackerrank.com/challenges/ctci-bfs-shortest-reach/problem
class Graph:
    class Node:
        def __init__(self):
            self.neighbors = []
    
    def __init__(self, size):
        self.edge_weight = 6
        self.nodes = []
        for i in range(size):
            new_node = self.Node()
            self.nodes.append(new_node)
    
    def connect(self, first, second):
        self.nodes[first].neighbors.append(second)
        self.nodes[second].neighbors.append(first)
    
    def find_all_distances(self, startId):
        queue = []
        queue.append(startId)
        
        distances = [-1] * len(self.nodes)
        distances[startId] = 0
        
        while (len(queue) > 0):
            node = queue.pop(0)
            for neighbor in self.nodes[node].neighbors:
                if distances[neighbor] == -1:
                    distances[neighbor] = distances[node] + self.edge_weight
                    queue.append(neighbor)
                    
        return distances


t = int(input())
for i in range(t):
    n, m = [int(value) for value in input().split()]
    graph = Graph(n)
    
    for i in range(m):
        x, y = [int(x) for x in input().split()]
        graph.connect(x - 1, y - 1)
    
    s = int(input()) - 1
    distances = graph.find_all_distances(s)
    
    for i in range(len(distances)):
        if i != s:
            print(distances[i], end=' ')
    print()
